### PR TITLE
iox-#1969 align iox expected to std expected and add convenience free functions

### LINKED ISSUE #1969: Better align 'iox::expected' with 'std::expected'
## Brief feature description

Currently the `iox::expected` differs from the newly introduced C++23 `std::expected`. To make it easier for developer already familiar with `std::expected` the custom implementation in iceoryx should be aligned where possible.

## Detailed information

In scope:
-  [x] ~~implicit conversion from `T`~~ -> we will not have implicit conversion; for safety code it is better to be explicit
-  [x] ~~default ctor creates an expected with a `T` if `T` has a default ctor~~ -> we will not have default ctors; for safety code it is better to be explicit
-  [x] introduction of `iox::unexpected`, `iox::unexpect_t` and `iox::unexpect`
-  [x] introduction of `iox::ok(...)` and `iox::err(...)` as convenience function due to missing [CTAD](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) in C++14
-  [x] rename and move of `iox::success` and `iox::error` to `iox::detail::ok` and `iox::detail::err`
-  [x] remove static `iox::expected<T, E>::create_value` and `iox::expected<T, E>::create_error` in favor of ctors with `iox::in_place` and `iox::unexpect`
-  [x] introduction of `iox::expected<T, E>::has_value`
-  [x] introduction of `iox::expected<T, E>::value` and deprecation of `iox::expected<T, E>::get_value`
-  [x] introduction of `iox::expected<T, E>::error` and deprecation of `iox::expected<T, E>::get_error`

Out of scope
- introduction of `transform`
- aligning of the semantics of `and_then` and `or_else`

### LINKED ISSUE #1976: `expected<void, Error>` is unusable due to `final`
## Required information

Since the `final` keyword was added to all `expected` classes the specialization `expected<void, Error>` is unusable since it inherit from them.
No one has noticed this until now so we have 2 solutions.
1. Remove `expected<void, Error>` specialization
2. Remove the `final` keyword.

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR aligns the `iox::expected` better to the `std::expected`
- remove `expected<E>` in favor of `expected<void, E>`
- replace `get_error` with `error`
- introduce `has_value`
- introduce `ok` and `error` free functions
- introduce `unexpect_t` helper struct for direct initialization

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] ~~Copyright owner are updated in the changed files~~
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes #1969 
- Closes #1976

### CODE REVIEW

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected_helper.hpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+  private:
+    static constexpr uint64_t VALUE_INDEX{0};
+    static constexpr uint64_t ERROR_INDEX{1};
+
+    using DummyValueType = bool;
+    static constexpr DummyValueType DUMMY_VALUE{true};
+
+    iox::variant<DummyValueType, ErrorType> data;
```
Comment (mossmaurice) [inline iceoryx_hoofs/vocabulary/include/iox/detail/expected_helper.hpp:218]: I think that should be enough, right?

```suggestion
    variant<DummyValueType, ErrorType> data;
```

File: iceoryx_hoofs/vocabulary/include/iox/expected.hpp [hunk 2]
```diff
@@ -18,254 +18,114 @@
...
-#include "iox/variant.hpp"
 
 #include <utility>
 
 namespace iox
 {
-/// @brief helper struct to create an expected which is signalling success more easily
-/// @param T type which the success helper class should contain
+template <typename T = void>
+using success = detail::ok<T>;
```
Comment (mossmaurice) [inline iceoryx_hoofs/vocabulary/include/iox/expected.hpp:30]: Shouldn't this be deprecated?

File: iceoryx_hoofs/vocabulary/include/iox/expected.hpp [hunk 3]
```diff
@@ -308,181 +176,202 @@ class IOX_NO_DISCARD expected<ValueType, ErrorType> final
...
 
-    /// @brief  returns a reference to the contained success value, if the expected
+    /// @brief  returns a const rvalue reference to the contained error value, if the expected
+    ///         does not contain an error the error handler is called
+    /// @return const rvalue reference to the internally contained error
+    const ErrorType&& error() const&& noexcept;
+
+    /// @copydoc expected::error()&
+    /// @deprecated use 'error' instead of 'get_error'
+    [[deprecated("Use 'error' instead of 'get_error'")]] ErrorType& get_error() & noexcept;
```
Comment (mossmaurice) [inline iceoryx_hoofs/vocabulary/include/iox/expected.hpp:246]: :heart:

**Review by mossmaurice — COMMENTED** [review]
LGTM content-wise, thanks! Sorry for being a smug here, but would you mind updating the year in the copyright headers? :sweat_smile:

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected_helper.hpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/expected_helper.hpp:218]: But it's not needed :)

**Review by mossmaurice — APPROVED** [review]
Approved under the condition that the copyright header is updated in a follow-up.

**Comment by elBoberido** [comment]
Sorry, all relevant copyright header are updated. A mass search and replace does not warrant an update of the copyright header.

**Pull request merged**